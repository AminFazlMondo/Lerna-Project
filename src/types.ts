import {javascript} from 'projen'

export interface LernaProjectOptions extends javascript.NodeProjectOptions {
  /**
   * (experimental) Use TypeScript for your projenrc file (`.projenrc.ts`).
   *
   * @default false
   * @experimental
   */
  readonly projenrcTs?: boolean;

  /**
   * Consolidate sub projects doc files
   *
   * @default false
   * @experimental
   */
  readonly docgen?: boolean;

  /**
   * (experimental) Docs directory.
   *
   * @default "docs"
   * @experimental
   */
  readonly docsDirectory?: string;

  /**
   * (experimental) Flag to run tasks only for the packages that has changes since last release
   *
   * @default false
   * @experimental
   */
  readonly sinceLastRelease?: boolean;

  /**
   * Whether or not to use Nx for task scheduling
   * https://lerna.js.org/docs/lerna-and-nx
   *
   * @default false
   */
  readonly useNx?: boolean;

  /**
   * Whether or not to use independent versioning for sub-projects
   * https://lerna.js.org/docs/features/version-and-publish#independent-mode
   *
   * @default false
   */
  readonly independentMode?: boolean;

}