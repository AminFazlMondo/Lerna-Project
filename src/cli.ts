import {Command} from 'commander'
import {copy, readdirSync, remove, existsSync, move} from 'fs-extra'

const program = new Command()

const distFolder = './dist'

program
  .command('clean-dist')
  .action(async () => {
    const entries = readdirSync(distFolder)
    await Promise.all(entries.map(f => remove(`${distFolder}/${f}`)))
  })

program
  .command('copy-dist <subProjectPath>')
  .action(async (subProjectPath) => {
    const subProjectDist = `./${subProjectPath}/dist`
    if (!existsSync(subProjectDist))
      return

    const entries = readdirSync(subProjectDist)
    await Promise.all(entries.map(f => copy(`${subProjectDist}/${f}`, `${distFolder}/${f}`, {recursive: true})))
  })


program
  .command('move-docs <parentDocsDirectory> <subProjectPath> <subProjectDocsDirectory>')
  .action(async (parentDocsDirectory, subProjectPath, subProjectDocsDirectory) => {
    await moveDocs(parentDocsDirectory, subProjectPath, subProjectDocsDirectory, false)
    await moveDocs(parentDocsDirectory, subProjectPath, 'API.md', true)
  })

async function moveDocs(parentDocsDirectory: string, subProjectPath: string, subPath: string, isFile: boolean): Promise<void> {
  const subProjectDocs = `./${subProjectPath}/${subPath}`
  if (!existsSync(subProjectDocs))
    return
  if (isFile) {
    await move(subProjectDocs, `./${parentDocsDirectory}/${subProjectPath}/${subPath}`)
  } else {
    const entries = readdirSync(subProjectDocs)
    await Promise.all(entries.map(f => move(`${subProjectDocs}/${f}`, `./${parentDocsDirectory}/${subProjectPath}/${f}`)))
  }
}

program.parse(process.argv)