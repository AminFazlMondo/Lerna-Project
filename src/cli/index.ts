export * from 'projen/lib/cli'