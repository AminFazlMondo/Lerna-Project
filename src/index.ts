import {JsonFile, javascript, Project, Tasks, SourceCode} from 'projen'
import {LernaProjectOptions} from './types'

export * from './types'

function getDocsDirectory(project: Project) {
  const result = Object.entries(project).find(([key]) => key === 'docsDirectory')
  return result?.[1].replace(/\/$/, '')
}

function getArtifactsDirectory(project: Project) {
  const result = Object.entries(project).find(([key]) => key === 'artifactsDirectory')
  return result?.[1]
}

const jsiiTaskPattern = /jsii-docgen -o (?<output>.+)$/i

function extractJsiiDocsOutput(tasks: Tasks): string | undefined {
  const jsiiDocStep = tasks.tryFind('docgen')?.steps.find(step => jsiiTaskPattern.test(step.exec ?? ''))
  if (!jsiiDocStep || !jsiiDocStep.exec)
    return

  const match = jsiiDocStep.exec.match(jsiiTaskPattern)
  return match?.groups?.output
}

const lockedTaskNames = ['build', 'upgrade', 'upgrade-projen']

export class LernaProject extends javascript.NodeProject {

  private subProjects: Record<string, Project>
  private projenrcTs: boolean

  readonly docsDirectory: string
  readonly docgen: boolean
  readonly sinceLastRelease: boolean
  readonly useNx: boolean
  readonly independentMode: boolean
  readonly useWorkspaces: boolean

  constructor(options: LernaProjectOptions) {
    super({
      ...options,
      jest: false,
    })

    this.addDevDeps('lerna-projen', 'lerna@5')

    if (options.projenrcTs)
      this.addDevDeps('ts-node', 'typescript')

    this.subProjects = {}
    this.docsDirectory = options.docsDirectory ?? 'docs'
    this.docgen = options.docgen ?? false
    this.sinceLastRelease = options.sinceLastRelease ?? false
    this.useNx = options.useNx ?? false
    this.projenrcTs = options.projenrcTs ?? false
    this.independentMode = options.independentMode ?? false
    this.useWorkspaces = options.useWorkspaces ?? false
  }

  addSubProject(subProject: Project) {
    const {outdir} = subProject

    if (!outdir || !outdir.includes(this.outdir))
      throw new Error('A sub project out dir should exists within the lerna package')

    const relativeOutDir = outdir.replace(`${this.outdir}/`, '')

    this.subProjects[relativeOutDir] = subProject
  }

  preSynthesize() {
    super.preSynthesize()
    const projenCommand = this.projenrcTs ? 'ts-node --skip-project .projenrc.ts' : 'node .projenrc.js'
    const {defaultTask} = this
    if (!defaultTask)
      throw new Error('Could not find default task')
    defaultTask.reset(projenCommand)
    this.packageTask.reset(`mkdir -p ${this.artifactsJavascriptDirectory}`)

    this.appendLernaCommands()

    this.preCompileTask.exec(`lerna-projen clean-dist ${this.artifactsDirectory}`)

    this.addCrossLinks()
    this.updateSubProjects()
    this.addDocumentsIndex()
  }

  private addCrossLinks() {
    const lernaConfig: any = {
      useNx: this.useNx,
      version: this.independentMode ? 'independent' : '0.0.0',
    }

    const packages = Object.keys(this.subProjects)

    if (this.useWorkspaces)
      this.package.addField('packages', packages)
    else
      lernaConfig.packages = packages


    new JsonFile(this, 'lerna.json', {
      obj: lernaConfig,
    })
  }

  private appendLernaCommands() {
    const upgradeTaskName = 'upgrade'
    const postUpgradeTaskName = 'post-upgrade'
    const postUpgradeTask = this.tasks.tryFind(postUpgradeTaskName)
    postUpgradeTask?.prependExec(this.getLernaCommand(upgradeTaskName, false))
    postUpgradeTask?.exec('npx projen')
    postUpgradeTask?.exec(this.getLernaCommand(postUpgradeTaskName))

    this.tasks.all
      .forEach(task => {
        if (lockedTaskNames.includes(task.name) || task.name === postUpgradeTaskName)
          return

        task.exec(this.getLernaCommand(task.name))
      })
  }

  private getLernaCommand(taskName: string, useSinceFlag = true) {
    const mainCommand = `lerna run ${taskName} --stream`
    const postCommand = useSinceFlag && this.sinceLastRelease ? ' --since $(git describe --abbrev=0 --tags --match "v*")' : ''
    return `${mainCommand}${postCommand}`
  }

  private updateSubProjects() {
    const bumpTask = this.tasks.tryFind('bump')
    const unbumpTask = this.tasks.tryFind('unbump')

    Object.entries(this.subProjects).forEach(([subProjectPath, subProject]) => {
      const subProjectDocsDirectory = getDocsDirectory(subProject)
      if (this.docgen && subProjectDocsDirectory)
        this.postCompileTask.exec(`lerna-projen move-docs ${this.docsDirectory} ${subProjectPath} ${subProjectDocsDirectory}`)

      const packageAllTask = subProject.tasks.tryFind('package-all')

      if (packageAllTask)
        subProject.packageTask.spawn(packageAllTask)

      const artifactsDirectory = getArtifactsDirectory(subProject)

      this.packageTask.exec(`lerna-projen copy-dist ${subProjectPath}/${artifactsDirectory} ${this.artifactsDirectory}`)

      subProject.defaultTask?.reset()

      const bumpEnvs = {
        OUTFILE: 'package.json',
        CHANGELOG: `${artifactsDirectory}/changelog.md`,
        BUMPFILE: `${artifactsDirectory}/version.txt`,
        RELEASETAG: `${artifactsDirectory}/releasetag.txt`,
        RELEASE_TAG_PREFIX: '',
      }

      if (bumpTask) {
        const subBumpTask = subProject.addTask(bumpTask.name, {
          description: bumpTask.description,
          condition: bumpTask.condition,
          env: bumpEnvs,
        })
        subBumpTask.builtin('release/bump-version')
      }

      if (unbumpTask) {
        const subBumpTask = subProject.addTask(unbumpTask.name, {
          description: unbumpTask.description,
          condition: unbumpTask.condition,
          env: bumpEnvs,
        })
        subBumpTask.builtin('release/reset-version')
      }
    })
  }

  private addDocumentsIndex() {
    if (!this.docgen)
      return

    const subProjectsDocs: Record<string, string> = {}

    const indexMarkdown = new SourceCode(this, `${this.docsDirectory}/index.md`)
    const readmeMarkdown = new SourceCode(this, `${this.docsDirectory}/README.md`)

    Object.entries(this.subProjects).forEach(([subProjectPath, subProject]) => {
      const subProjectDocsDirectory = getDocsDirectory(subProject)
      if (!subProjectDocsDirectory)
        return

      const jsiiDocsOutput = extractJsiiDocsOutput(subProject.tasks)
      if (jsiiDocsOutput) {
        const path = `${subProjectPath}/${jsiiDocsOutput}`
        indexMarkdown.line(`- ## [${subProject.name}](${path})`)
        subProjectsDocs[subProject.name] = path
      } else {
        subProjectsDocs[subProject.name] = `${subProjectPath}/index.html`
      }
      readmeMarkdown.line(`- ## [${subProject.name}](../${subProjectPath}/README.md)`)
    })

    const indexHtml = new SourceCode(this, `${this.docsDirectory}/index.html`)
    indexHtml.line('<!DOCTYPE html>')
    indexHtml.line('<html>')
    indexHtml.open('<body>')
    indexHtml.open('<ul>')
    Object.entries(subProjectsDocs).forEach(([name, path]) => {
      indexHtml.open('<li>')
      indexHtml.open(`<a href="${path}">`)
      indexHtml.line(name)
      indexHtml.close('</a>')
      indexHtml.close('</li>')
    })
    indexHtml.close('</ul>')
    indexHtml.close('</body>')
    indexHtml.line('</html>')
    indexHtml.line()

    this.gitattributes.addAttributes(`/${this.docsDirectory}/**`, 'linguist-generated')
  }
}