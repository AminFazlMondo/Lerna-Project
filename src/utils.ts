import {NodeProject} from 'projen/lib/javascript'

/**
 * Adds dependency to the project for NX task runner
 * see https://nx.dev/reference/project-configuration#dependson
 *
 * @param project The Project to add dependency to
 * @param taskName The task name that is dependent on another tasks
 * @param dependsOnTaskName The task name that is dependent on in other projects
 * @param dependsOn The packages that source project is dependent on
 */
export function addNxTaskDependency(project: NodeProject, taskName: string, dependsOnTaskName: string, ...dependsOn: NodeProject[]): void {
  project.package.addField('nx', {
    targets: {
      [taskName]: {
        dependsOn: [
          {
            projects: dependsOn.map(d => d.package.packageName),
            target: dependsOnTaskName,
          },
        ],
      },
    },
  })
}