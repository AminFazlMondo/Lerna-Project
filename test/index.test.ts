import * as fs from 'fs'
import * as os from 'os'
import * as path from 'path'
import {removeSync} from 'fs-extra'
import {LogLevel, NodeProject, TypeScriptProject} from 'projen'
import {LernaProject} from '../src'

const autoRemove = new Set<string>()

afterAll((done) => {
  Array.from(autoRemove).forEach(dir => removeSync(dir))
  done()
})

function mkdtemp() {
  const tmpdir = fs.mkdtempSync(path.join(os.tmpdir(), 'projen-test-'))
  autoRemove.add(tmpdir)
  return tmpdir
}

interface SynthOutput {
  'lerna.json': any;
  'tasks.json': any;
}

function readJson(filePath: string): any {
  const content = fs.readFileSync(filePath, 'utf-8')
  return JSON.parse(content)
}

function captureSynth(project: LernaProject): SynthOutput {
  process.env.PROJEN_DISABLE_POST = 'true'
  project.synth()
  const {outdir} = project
  return {
    'lerna.json': readJson(path.join(outdir, 'lerna.json')),
    'tasks.json': readJson(path.join(outdir, '.projen', 'tasks.json')),
  }
}
interface GenerateProjectsParams {
  /**
   * @default false
   */
  docgen?: boolean;

  /**
   * @default true
   */
  subProjectHasDocs?: boolean;

  /**
   * @default false
   */
  sinceLastRelease?: boolean;

  /**
   * @default false
   */
  projenrcTs?: boolean;
}
// eslint-disable-next-line max-len
function generateProjects(parentDocsFolder: string, subProjectDirectory: string, params: GenerateProjectsParams = {}): LernaProject {
  const parentDirectory = mkdtemp()
  const parentProject = new LernaProject({
    name: 'test',
    outdir: parentDirectory,
    defaultReleaseBranch: 'test',
    logging: {
      level: LogLevel.OFF,
    },
    docsDirectory: parentDocsFolder,
    docgen: params.docgen ?? false,
    sinceLastRelease: params.sinceLastRelease ?? false,
    projenrcTs: params.projenrcTs ?? false,
  })

  const SubProjectType = (params.subProjectHasDocs ?? true) ? TypeScriptProject : NodeProject
  const subProject = new SubProjectType({
    name: 'test-sub-project',
    parent: parentProject,
    outdir: subProjectDirectory,
    defaultReleaseBranch: 'test',
    logging: {
      level: LogLevel.OFF,
    },
  })
  parentProject.addSubProject(subProject)
  return parentProject
}

const parentDocsFolder = 'stub-docs'
const subProjectDirectory = 'packages/test-sub-project'
const expectedDocsCommand = `lerna-projen move-docs ${parentDocsFolder} ${subProjectDirectory} docs`

describe('Happy Path', () => {
  let parentProject: LernaProject

  beforeEach(() => {
    parentProject = generateProjects(parentDocsFolder, subProjectDirectory)
  })

  test('lerna file', () => {
    const output = captureSynth(parentProject)
    expect(output['lerna.json']).toMatchObject({
      packages: [subProjectDirectory],
      version: '4.0.0',
    })
  })

  describe('tasks', () => {
    test('default', () => {
      const output = captureSynth(parentProject)
      expect(output['tasks.json']).toEqual(
        expect.objectContaining({
          tasks: expect.objectContaining({
            default: expect.objectContaining({
              steps: expect.not.arrayContaining([{
                exec: 'lerna run test --stream',
              }]),
            }),
          }),
        }),
      )

    })

    test('test', () => {
      const output = captureSynth(parentProject)
      expect(output['tasks.json']).toEqual(
        expect.objectContaining({
          tasks: expect.objectContaining({
            test: expect.objectContaining({
              steps: expect.arrayContaining([{
                exec: 'lerna run test --stream',
              }]),
            }),
          }),
        }),
      )

    })

    test('build', () => {
      const output = captureSynth(parentProject)
      expect(output['tasks.json']).toEqual(
        expect.objectContaining({
          tasks: expect.objectContaining({
            build: expect.objectContaining({
              steps: expect.not.arrayContaining([
                {
                  exec: 'lerna run build --stream',
                },
                {
                  exec: 'lerna-projen clean-dist',
                },
                {
                  exec: `lerna-projen copy-dist ${subProjectDirectory}`,
                },
              ]),
            }),
          }),
        }),
      )

    })

    test('pre-compile', () => {
      const output = captureSynth(parentProject)
      expect(output['tasks.json']).toEqual(
        expect.objectContaining({
          tasks: expect.objectContaining({
            ['pre-compile']: expect.objectContaining({
              steps: expect.arrayContaining([
                {
                  exec: 'lerna-projen clean-dist',
                },
                {
                  exec: 'lerna run pre-compile --stream',
                },
              ]),
            }),
          }),
        }),
      )

    })

    test('package', () => {
      const output = captureSynth(parentProject)
      expect(output['tasks.json']).toEqual(
        expect.objectContaining({
          tasks: expect.objectContaining({
            package: expect.objectContaining({
              steps: expect.arrayContaining([
                {
                  exec: 'lerna run package --stream',
                },
                {
                  exec: `lerna-projen copy-dist ${subProjectDirectory}`,
                },
              ]),
            }),
          }),
        }),
      )

    })

    test('should not include docs tasks by default', ()=> {
      const output = captureSynth(parentProject)
      expect(output['tasks.json']).toEqual(
        expect.objectContaining({
          tasks: expect.objectContaining({
            build: expect.objectContaining({
              steps: expect.not.arrayContaining([
                {
                  exec: expectedDocsCommand,
                },
              ]),
            }),
          }),
        }),
      )

    })

  })
})


describe('docgen set to true', () => {
  test('sub project has docs', () => {
    const parentProject = generateProjects(parentDocsFolder, subProjectDirectory, {docgen: true})
    const output = captureSynth(parentProject)
    expect(output['tasks.json']).toEqual(
      expect.objectContaining({
        tasks: expect.objectContaining({
          ['post-compile']: expect.objectContaining({
            steps: expect.arrayContaining([
              {
                exec: expectedDocsCommand,
              },
            ]),
          }),
        }),
      }),
    )
  })

  test('sub project does not have docs', () => {
    const parentProject = generateProjects(parentDocsFolder, subProjectDirectory, {docgen: true, subProjectHasDocs: false})
    const output = captureSynth(parentProject)
    expect(output['tasks.json']).toEqual(
      expect.objectContaining({
        tasks: expect.objectContaining({
          build: expect.objectContaining({
            steps: expect.not.arrayContaining([
              {
                exec: expectedDocsCommand,
              },
            ]),
          }),
        }),
      }),
    )
  })
})

describe('since last release', () => {
  test('should include since filter', () => {
    const parentProject = generateProjects(parentDocsFolder, subProjectDirectory, {sinceLastRelease: true})
    const output = captureSynth(parentProject)
    expect(output['tasks.json']).toEqual(
      expect.objectContaining({
        tasks: expect.objectContaining({
          compile: expect.objectContaining({
            steps: expect.arrayContaining([
              {
                exec: 'lerna run compile --stream --since $(git describe --abbrev=0 --tags --match "v*")',
              },
            ]),
          }),
        }),
      }),
    )
  })
})

describe('typescript projenrc file', () => {
  test('should use the projenrc.ts file', () => {
    const parentProject = generateProjects(parentDocsFolder, subProjectDirectory, {projenrcTs: true})
    const output = captureSynth(parentProject)
    expect(output['tasks.json']).toEqual(
      expect.objectContaining({
        tasks: expect.objectContaining({
          default: expect.objectContaining({
            steps: expect.arrayContaining([
              {
                exec: 'ts-node --skip-project .projenrc.ts',
              },
            ]),
          }),
        }),
      }),
    )
  })
})

describe('Unhappy Path', () => {

  describe('output directory', () => {

    test('Should fail if sub project does not have output directory', () => {
      const parentDirectory = mkdtemp()
      const project = new LernaProject({
        name: 'test',
        outdir: parentDirectory,
        defaultReleaseBranch: 'test',
        logging: {
          level: LogLevel.OFF,
        },
      })

      const subProject = new TypeScriptProject({
        name: 'test-sub-project',
        defaultReleaseBranch: 'test',
        logging: {
          level: LogLevel.OFF,
        },
      })
      expect(() => project.addSubProject(subProject)).toThrow('A sub project out dir should exists within the lerna package')
    })

    test('Should fail if sub project does not have output directory outside of parent root', () => {
      const parentDirectory = mkdtemp()
      const project = new LernaProject({
        name: 'test',
        outdir: parentDirectory,
        defaultReleaseBranch: 'test',
        logging: {
          level: LogLevel.OFF,
        },
      })

      const subProject = new TypeScriptProject({
        name: 'test-sub-project',
        outdir: mkdtemp(),
        defaultReleaseBranch: 'test',
        logging: {
          level: LogLevel.OFF,
        },
      })
      expect(() => project.addSubProject(subProject)).toThrow('A sub project out dir should exists within the lerna package')
    })

  })

})