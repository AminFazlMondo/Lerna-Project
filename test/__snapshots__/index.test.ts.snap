// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Happy Path for Jsii sub project sub project has docs docs index 1`] = `
"<!DOCTYPE html>
<html>
<body>
  <ul>
    <li>
      <a href=\\"packages/test-sub-project/API.md\\">
        test-sub-project
      </a>
    </li>
  </ul>
</body>
</html>
"
`;

exports[`Happy Path for Jsii sub project sub project has docs docs index 2`] = `"- ## [test-sub-project](packages/test-sub-project/API.md)"`;

exports[`docgen set to true multiple sub-projects 1`] = `
"<!DOCTYPE html>
<html>
<body>
  <ul>
    <li>
      <a href=\\"packages/test-sub-project-1/API.md\\">
        test-sub-project-1
      </a>
    </li>
    <li>
      <a href=\\"packages/test-sub-project-2/index.html\\">
        test-sub-project-2
      </a>
    </li>
    <li>
      <a href=\\"packages/test-sub-project-4/API.md\\">
        test-sub-project-4
      </a>
    </li>
  </ul>
</body>
</html>
"
`;

exports[`docgen set to true multiple sub-projects 2`] = `
"- ## [test-sub-project-1](packages/test-sub-project-1/API.md)
- ## [test-sub-project-4](packages/test-sub-project-4/API.md)"
`;

exports[`docgen set to true sub project has docs docs index 1`] = `
"<!DOCTYPE html>
<html>
<body>
  <ul>
    <li>
      <a href=\\"packages/test-sub-project/index.html\\">
        test-sub-project
      </a>
    </li>
  </ul>
</body>
</html>
"
`;

exports[`docgen set to true sub project has docs docs index 2`] = `""`;
